six/common/aggregator.c

```diff
@@ -84,42 +137,35 @@ static PyObject *submit_metric(PyObject *self, PyObject *args) {
 
     // Python call: aggregator.submit_metric(self, check_id, aggregator.metric_type.GAUGE, name, value, tags, hostname)
     if (!PyArg_ParseTuple(args, "OsisfOs", &check, &check_id, &mt, &name, &value, &py_tags, &hostname)) {
-        goto done;
+        PyErr_SetString(PyExc_TypeError, "wrong parameters type");
+        goto error;
     }
 
-    int len = PySequence_Length(py_tags);
-    if (len) {
-        py_tags_list = PySequence_Fast(py_tags, err); // new reference
-        if (py_tags_list == NULL) {
-            goto done;
-        }
+    if ((tags = py_tag_to_c_tags(py_tags)) == NULL)
+        goto error;
 
-        tags = malloc(len * sizeof(char *));
-        int i;
-        for (i = 0; i < len; i++) {
-            // `item` is borrowed, no need to decref
-            PyObject *item = PySequence_Fast_GET_ITEM(py_tags_list, i);
-            tags[i] = as_string(item);
-        }
-    }
+    cb_submit_metric(check_id, mt, name, value, tags, hostname);
 
-    cb_submit_metric(check_id, mt, name, value, tags, len, hostname);
+    free_tags(tags);
 
-done:
-    free(err);
-    Py_XDECREF(py_tags_list);
+    PyGILState_Release(gstate);
     Py_RETURN_NONE;
+
+error:
+    PyGILState_Release(gstate);
+    // we need to return NULL to raise the exception set by PyErr_SetString
+    return NULL;
 }
 
 static PyObject *submit_service_check(PyObject *self, PyObject *args) {
     if (cb_submit_service_check == NULL) {
         Py_RETURN_NONE;
     }
 
+    PyGILState_STATE gstate = PyGILState_Ensure();
```

Comment:
nit: maybe worth it adding a comment like "since we can potentially raise a Python exception we need to lock the GIL" but we can review these things later in a specific PR